### PR TITLE
Serialize all models into cluster metadata

### PR DESCRIPTION
**Pull request by ryanbogan**
### Description
This PR removes a check that only serialized models into cluster metadata when they were in the "created" state.  This means that any model in "training" or "failed" states could only be accessed through blocking transport requests.  This PR also minimizes the transport calls in `TrainingJobClusterStateListener`, instead accessing model metadata from the cluster metadata.

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 1]
```diff
@@ -330,13 +330,7 @@ private void putInternal(Model model, ActionListener<IndexResponse> listener, Do
...
             }, listener::onFailure);
 
-            // After the model is indexed, update metadata only if the model is in CREATED state
-            ActionListener<IndexResponse> onIndexListener;
-            if (ModelState.CREATED.equals(model.getModelMetadata().getState())) {
-                onIndexListener = getUpdateModelMetadataListener(model.getModelMetadata(), onMetaListener);
-            } else {
-                onIndexListener = onMetaListener;
-            }
+            ActionListener<IndexResponse> onIndexListener = getUpdateModelMetadataListener(model.getModelMetadata(), onMetaListener);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelDao.java:333]: Want to discuss a little bit here some thoughts:

So before, the logic was that the model index is the only source of truth until the model is actually created. Then the model-metadata is a source of truth as well.

Now, we are saying that the model metadata will always lag behind the model system index (i.e. on any call to change model state, we will first persist in system index and then update in cluster state). This is going to open up to the possibility that model index and cluster state fall out of sync on failure. We may need to think about how we handle different scenarios. 

I think a model's state can be updated in one of the following ways:
1. Training process (this will be in sync - single thread, synchronous process)
2. Model deleted (we block the model delete if the model is not in the created or error state - we should confirm that we check this from cluster metadata before blocking)
3. Node drop (if a node drops, the offending node will know not to change anything in the state correct?)

Are there other cases you can think of that might be of concern?

**Review by jmazanec15 — COMMENTED** [review]
We need to update here as well: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java#L48-L52

**Comment by jmazanec15** [comment]
@ryanbogan and check from here: https://github.com/opensearch-project/k-NN/blob/7a144b84f8af1e178054d34cd716ce139209f1b5/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L213-L221

**Comment by jmazanec15** [comment]
and here: https://github.com/opensearch-project/k-NN/blob/7a144b84f8af1e178054d34cd716ce139209f1b5/src/main/java/org/opensearch/knn/index/IndexUtil.java#L25

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 1 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:333]: @jmazanec15 The goal of this was to treat the cluster metadata as the main source of truth in case of node drops I believe

**Comment by ryanbogan** [comment]
> @ryanbogan and check from here:
> 
> https://github.com/opensearch-project/k-NN/blob/7a144b84f8af1e178054d34cd716ce139209f1b5/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L213-L221

I'll add a check to get the state

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 1 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:344]: I wrote a temporary integration test trying to create a race condition between a model finishing training and entering the created state in ModelDao and then deleting the model immediately.  All behavior was as expected and the model was deleted from the cluster metadata as well.

File: src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java [hunk 2]
```diff
@@ -109,10 +109,9 @@ protected void updateModelsNewCluster() throws IOException, InterruptedException
         if (modelDao.isCreated()) {
             List<String> modelIds = searchModelIds();
             for (String modelId : modelIds) {
-                Model model = modelDao.get(modelId);
-                ModelMetadata modelMetadata = model.getModelMetadata();
+                ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java:112]: Actually, I'm not sure if there is a way to make it backwards compatible, since the old models wouldn't be in the cluster metadata

File: jni/CMakeLists.txt [hunk 3]
```diff
@@ -129,7 +129,7 @@ if (${CONFIG_FAISS} STREQUAL ON OR ${CONFIG_ALL} STREQUAL ON OR ${CONFIG_TEST} S
         set(SIMD_ENABLED true)   # set default value as true if the argument is not set
     endif()
 
-    if(${CMAKE_SYSTEM_NAME} STREQUAL Windows OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64" OR NOT ${SIMD_ENABLED})
+    if(${CMAKE_SYSTEM_NAME} STREQUAL Windows OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64" OR NOT ${SIMD_ENABLED})
```
Comment (martin-gaievski) [inline jni/CMakeLists.txt:132]: if that's a separate fix unrelated to the logic on this feature I'd rather put into a separate PR mainly because it's easy to track and backport and it can be merged earlier to unblock other PRs (as this PR may take some time to merge)

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:344]: @ryanbogan can you push to another branch and link to the test? I want to take a quick look.

**Comment by ryanbogan** [comment]
Windows failures are unrelated to this PR

File: src/main/java/org/opensearch/knn/index/IndexUtil.java [hunk 4]
```diff
@@ -197,7 +198,7 @@ public static ValidationException validateKnnField(
             }
 
             ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
-            if (modelMetadata == null) {
+            if (!ModelUtil.isModelPresent(modelMetadata) || !ModelUtil.isModelCreated(modelMetadata)) {
                 exception.addValidationError(String.format("Model \"%s\" for field \"%s\" does not exist.", modelId, field));
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/IndexUtil.java:202]: update error message here

File: src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java [hunk 5]
```diff
@@ -50,7 +51,7 @@ protected void parseCreateField(ParseContext context) throws IOException {
         // model when ingestion starts.
         ModelMetadata modelMetadata = this.modelDao.getMetadata(modelId);
 
-        if (modelMetadata == null) {
+        if (!ModelUtil.isModelPresent(modelMetadata) || !ModelUtil.isModelCreated(modelMetadata)) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java:54]: Cant this just be !ModelUtil.isModelCreated? Same as above

File: src/main/java/org/opensearch/knn/indices/ModelUtil.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.indices;
+
+public class ModelUtil {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelUtil.java:14]: add javadocs for this class

File: src/main/java/org/opensearch/knn/indices/ModelUtil.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    public static boolean isModelPresent(ModelMetadata modelMetadata) {
+        return modelMetadata != null;
+    }
+
+    public static boolean isModelCreated(ModelMetadata modelMetadata) {
+        return modelMetadata.getState().equals(ModelState.CREATED);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelUtil.java:21]: add null check on modelMetadata

**Review by jmazanec15 — COMMENTED** [review]
a few minor comments. overall looks good

File: src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java (hunk 5 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java:54]: We need to check if it's null first, otherwise getState will throw a null pointer exception instead of the exception we want to throw

Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java:54]: Unless we add the null check like your comment below says, then that would work!

File: src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java [hunk 8]
```diff
@@ -112,7 +112,7 @@ protected void updateModelsNewCluster() throws IOException, InterruptedException
                 Model model = modelDao.get(modelId);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java:112]: We can remove this line.

Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java:112]: Per our discussion offline, I changed the line and refactored the code to ensure backwards compatibility.

**Review by jmazanec15 — APPROVED** [review]
LGTM thanks!

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 9]
```diff
@@ -213,8 +214,8 @@ private Map<Integer, Float> doANNSearch(final LeafReaderContext context, final B
         String modelId = fieldInfo.getAttribute(MODEL_ID);
         if (modelId != null) {
             ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
-            if (modelMetadata == null) {
-                throw new RuntimeException("Model \"" + modelId + "\" does not exist.");
+            if (!ModelUtil.isModelCreated(modelMetadata)) {
+                throw new RuntimeException("Model \"" + modelId + "\" is not created.");
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:218]: nit: please use String.format for string concatenation

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:218]: @martin-gaievski this comes up a couple times because we do not do it for a lot of existing code. Do you know if its possible to add something to spotless to automatically fix this?

Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:218]: it should be possible, we did some custom template for spotless in neural-search, please check this PR https://github.com/opensearch-project/neural-search/pull/515

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:218]: @martin-gaievski is there a comment for it or file? Im not seeing it in formatting
